pkg/core/resources/apis/meshservice/generate/generator.go

```diff
@@ -175,22 +180,51 @@ func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*cor
 			log.Info("Port conflict for a kuma.io/service tag, ports must be identical across Dataplane inbounds for a given kuma.io/service", "dps", dps)
 		}
 		delete(meshservicesByName, meshService.GetMeta().GetName())
-		if newMeshService != nil && servicesDiffer(meshService.Spec, newMeshService) {
+		gracePeriodStartedAtText, hasGracePeriodLabel := meshService.GetMeta().GetLabels()[gracePeriodStartedLabel]
+
+		if newMeshService != nil && (servicesDiffer(meshService.Spec, newMeshService) || hasGracePeriodLabel) {
 			meta := meshService.GetMeta()
 			meshService = meshservice_api.NewMeshServiceResource()
 			meshService.Meta = meta
 			meshService.Spec = newMeshService
-			if err := g.resManager.Update(ctx, meshService); err != nil {
+
+			// Unset the grace period by deleting the label
+			newLabels := maps.Clone(meshService.GetMeta().GetLabels())
+			delete(newLabels, gracePeriodStartedLabel)
```

Comment:
here it would be nice to log something like `mesh service previously marked as ready to delete won't be deleted because there is a healthy dataplane proxy`

Reply:
done